File: Python/Heap/BJAlgo_1766.py
from sys import stdin
import heapq

n, m = map(int, stdin.readline().split())
array = [[] for _ in range(n + 1)]  # 해당 노드와 연결된 노드 담기
indegree = [0] * (n + 1)  # 진입 차수

heap = []

for _ in range(m):
    x, y = map(int, stdin.readline().split())
    array[x].append(y)  # x -> y 관계 연결
    indegree[y] += 1  # y 로 들어오는 진입 차수 증가

for i in range(1, n + 1):
    if indegree[i] == 0:  # 진입 차수가 0인 경우 heap 에 추가
        heapq.heappush(heap, i)

result = []
while heap:
    data = heapq.heappop(heap)
    result.append(data)
    for d in array[data]:  # 꺼낸 데이터와 연결된 데이터들을 돌면서
        indegree[d] -= 1   # 진입 차수 줄이기
        if indegree[d] == 0:  # 진입 차수가 0인 경우 heap 에 추가
            heapq.heappush(heap, d)

for i in result:
    print(i, end=" ")

File: Java/Heap/BJAlgo_1766.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;

public class BJAlgo_1766 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] nm = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        List<ArrayList<Integer>> array = new ArrayList<>();
        for (int i = 0; i < nm[0] + 1; i++) {
            array.add(new ArrayList<>());
        }
        int[] indegree = new int[nm[0] + 1];

        for (int i = 0; i < nm[1]; i++) {
            int[] xy = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            array.get(xy[0]).add(xy[1]);
            indegree[xy[1]] += 1;
        }

        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for (int i = 1; i < nm[0] + 1; i++) {
            if (indegree[i] == 0) {
                heap.add(i);
            }
        }

        StringBuilder result = new StringBuilder();
        while (!heap.isEmpty()) {
            int data = heap.poll();
            result.append(data).append(" ");

            for (int n : array.get(data)) {
                indegree[n] -= 1;
                if (indegree[n] == 0) {
                    heap.add(n);
                }
            }
        }

        System.out.println(result);
    }
}
